civicrm_entity.module

```diff
@@ -179,3 +179,124 @@ function _civicrm_entity_get_entity_type_from_elements($elements) {
 
   return NULL;
 }
+
+/**
+ * Implements hook_civicrm_pre().
+ */
+function civicrm_entity_civicrm_pre($op, $objectName, $id, &$params) {
+  $operations = ['create', 'edit', 'delete', 'restore'];
+  if (!in_array($op, $operations)) {
+    return;
+  }
+
+  $entityType = SupportedEntities::getEntityType($objectName);
+
+  // Check valid entity type.
+  $config = \Drupal::config('civicrm_entity.settings');
+  $enabled_entity_types = $config->get('enabled_entity_types') ?: [];
+  if (!in_array($entityType, $enabled_entity_types)) {
+    return;
```

Comment:
I'm not 100% on this, but I think we want the hooks to fire even if the entity isn't "enabled". When Matt and I originally discussed the enabled/disabled thing we came to the conclusion that it should just hide/show the entity type in the UI, because making the underlying entity stuff behave differently makes for a very weird, undeclarable dependency. But this is a little different than the issue we originally discussed...

Reply:
My main concern was that we just assume a Drupal entity type exists, then start using the storage service, etc. There are some things that are (or at least via the API appear to be) Civi objects but aren't defined for use in this module's `SupportedEntities` class.

Perhaps just call `SupportedEntities` to see if it's defined there?

Reply:
@dsnopek your last comment would be a departure from the D7 functionality, where if an entity is not enabled, then it would NOT engage any Drupal systems. The original idea behind that was basically a performance and permissions access issue, because over 40 entity types are exposed to Drupal. If an entity was disabled, then it would not be included in civicrm_entity_entity_info, and would be unavailable for any Drupal API integration...You do have a good point though about the undeclared dependency issue though...because for instance if a Rule had been created to respond to a certain entity CRUD operation, and then later an admin disables an entity type, then you have a broken Rule left hanging...Or if a custom module manipulated a civicrm entity type and then all of a sudden it's not exposed anymore you could conceivably get a broken site...Basically it was assumed that it is the module developers or site admins responsibility to keep track of or manage what is enabled. What I did for some of the submodules of the D7 version was on install/enable, if a entity type was required, it would enable the entity, and disable the checkbox element on the settings form so that required entity types could NOT be disabled via the UI, while the submodule was enabled. Can you lay out the case a little more for why the Drupal hooks should be called even if an entity type is disabled? Maybe there are other reasons, or D8 specific reasons I am not thinking of as well? I'll think about it some more too, you are right that it is a complex dependency to understand for people developing modules that use CiviCRM Entity. I'm also seeing how people could have trouble with broken Views if some admin disabled an entity type, and Views integration for that entity type goes away.

Reply:
Yeah, I think this is something we need to do differently for D8. In D8, there's a number of places where you're required to declare the dependencies something has in order for everything to work correctly. For example, every config object in CMI needs to say what it depends on. If a bit of config needs a particular entity type from civicrm_entity, it can depend on the civicrm_entity module, but it can't depend on civicrm_entity plus that the fact that the entity type is enabled, because that is just a specific property within another config object (we can declare that).

Side note: if civicrm_entity types were enabled by creating an entity, this would be different, because a config object can depend on an entity, but do we really want a whole entity type just to enable entity types? That sounds a little crazy.

So, the simple solution is to say the entity types always exist, but we then we try to hide the disabled ones to prevent exploding the UI.

Reply:
Ok thanks for your patience and explanation, let's not get crazy by using an entity type to enable entity types :) I think what we could do in the future is offer developers a way to exclude/include supported entity types. We could add an _alter() hook in getInfo() method of the SupportedEntities class....or we turn that class into a service and people could use service decorators, or maybe supported entities becomes a plugin type. It will become an issue down the road when people want to be able to integrate their custom entity types, and alter default properties of supported entities.

Reply:
Re @kerasai: 

> Perhaps just call `SupportedEntities` to see if it's defined there?

Yeah, I think that's the way to go.

Re @jackrabbithanna:

> I think what we could do in the future is offer developers a way to exclude/include supported entity types. We could add an _alter() hook in getInfo() method of the SupportedEntities class....or we turn that class into a service and people could use service decorators, or maybe supported entities becomes a plugin type.

I concur, we need some mechanism for this.

On one of Matt's earlier PR's I put forward the idea that the `SupportedEntities::getInfo()` information gets transferred straight to the entity types and then we look only at the entity types, so that other modules could alter the data (there's already an alter hook for entity types).

But having our own alter hook or using service decorators or maybe an event, all sound like fine ideas too. I'm kind of skeptical of using a plugin type, since the entity types can't be declared arbitrarily (the civicrm entity type needs to already exist), but there is some precedent in contrib for using plugin types like that (ie. more like 'handlers' than plugins).

Reply:
@kerasai For sure checking in SupportedEntities is the way to go here, since we will be standardizing this way for determining what entities are exposed as Drupal entities. If anywhere else in the code base, the check should be made from there too. @dsnopek youll think though excepting manage fields forms, manage display forms, or drupal native pages / forms?  . I can imagine developers wanting to at least add fields and manage them, if CiviCRM data was used by the REST API, or for other programmatic uses, say in these hooks....perhaps only Manage Display and the view page / edit forms should be excluded when disabled via the config page.